### PR TITLE
Remove older layer metadata format support code

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
The PR enforces current newest `index_part.json` format in the type system (version `1`), not allowing any previous forms of it, that were used in the past.
Similarly, the code to mitigate the https://github.com/neondatabase/neon/issues/3024 issue is now also removed.

Current code does not produce old formats and extra files in the index_part.json, in the future we will be able to use https://github.com/neondatabase/aversion or other approach to make version transitions more explicit.

Future planned transitions:
* adding `hash` and `hash_type` fields in the layer metadata for future implementation of bitrot and other network corruption checks
* merging `layer_metadata` and `timeline_layers` fields, to ensure on the type level, that all layers have their metadata calculated
* `metadata_bytes` extraction: at least, its version is very important to have visible

---------------

This transition is a breaking change, but it's possible due to all prod S3 data migrated to the newest format already.
I've processed ~900_000 timelines
and all of them:

* have `layer_metadata` and `timeline_layers` to contain the same set of layers
* all layers having their size in the metadata
* no *`.old`, `metadata_backup` or other non-layer files were found

Also, there are abnormalities in these jsons:
* ~56_000 timelines have 0 layers in them, which seems odd, even though does not contradict and fully compatible with the breaking changes above
* ~490_000 timelines still have `missing_layers` field in them (empty), which means they were not updated since the latest json file format update. Again, does not contradict and fully compatible with the breaking changes above: serde_json ignores "extra" fields when deserializing
* 2 timelines still have version 0 in their index_part.json, which means they were not updated for a long while

Since we do not delete S3 timeline files on project deletion (see https://github.com/neondatabase/neon/issues/2606 for more details on the deletion),all these abnormalities could be explained by their projects were deleted in console.
In practice, only two timelines with version 0 are incompatible with these changes, but these timelines were not used for a long time and most probably could be deleted.

### CODE REVIEW

**Review by koivunej — COMMENTED** [review]
Few early comments, also asked about the staging compatibility. We probably need a plan for that, which does not have to be full compatibility because there is a lot of cruft. Perhaps we could however somehow get an update for all non-removed tenants..? Though, might make sense to do that together with something which starts to mark tenants with missing layer files as broken.

> Current code does not produce old formats and extra files in the index_part.json, in the future we will be able to use https://github.com/neondatabase/aversion or other approach to make version transitions more explicit.

The soft json approach has worked but apparently it didn't save so much code (generic parameter from IndexPartImpl which this PR removes) ... Though looking at it, I am not sure what did that solve.

Another note on `aversion`, it expects a infallible conversion path from type to type, which I think requires injection of "additional data", like this `file_size` migration without an `Option<u64>` would had required. I don't think that has any benefit over json softness.

**Comment by hlinnaka** [comment]
> Since we do not delete S3 timeline files on project deletion (see #2606 for more details on the deletion),all these abnormalities could be explained by their projects were deleted in console.
> In practice, only two timelines with version 0 are incompatible with these changes, but these timelines were not used for a long time and most probably could be deleted.

They are probably deleted projects, but we need to verify that. Let's cross-check the list of these abnormalities with list of deleted projects, and ignore all the deleted ones.

File: pageserver/src/tenant/remote_timeline_client/index.rs [hunk 1]
```diff
@@ -263,8 +162,7 @@ mod tests {
...
         };
 
-        let part: IndexPartUnclean = serde_json::from_str(example).unwrap();
-        let part = part.remove_unclean_layer_file_names();
+        let part: IndexPart = serde_json::from_str(example).unwrap();
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/tenant/remote_timeline_client/index.rs:165]: Thanks, removed the annotation and the test in https://github.com/neondatabase/neon/pull/3854/commits/ce3aa1b928de12d146025e67f8f70e7fb55d3373, added a test case for empty files json.

File: pageserver/src/tenant/upload_queue.rs [hunk 2]
```diff
@@ -127,12 +128,21 @@ impl UploadQueue {
...
-            files.insert(layer_name.to_owned(), layer_metadata);
+            {
+                Some(layer_metadata) => {
+                    files.insert(layer_name.to_owned(), layer_metadata);
+                }
+                None => {
+                    error!(
+                        "No remote layer metadata found for layer {}",
+                        layer_name.file_name()
+                    );
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/tenant/upload_queue.rs:143]: Changed that to be `anyhow::bail`, thank you.

**Pull request closed without merging**